src/toil/test/server/wes_cwl_runner.py

```diff
@@ -0,0 +1,165 @@
+    --clean=always \
+    --logDebug
+```
+"""
+
+logger = logging.getLogger(__name__)
+
+
+def get_deps_from_cwltool(cwl_file, input_file, input_deps: bool = False):
+    """
+    Return a list of dependencies of the given workflow from cwltool.
+    """
+    if input_deps and not input_file:
+        raise ValueError("Must provide an input_file if input_deps = True.")
+
+    option = '--print-input-deps' if input_deps else '--print-deps'
+
+    args = ['cwltool', option, '--relative-deps', 'cwd', cwl_file]
+    if input_file:
+        args.append(input_file)
+
+    p = subprocess.run(args=args, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL)
+
+    result = p.stdout.decode()
+    if not result:
+        return []
+
+    json_result: Dict[str, Any] = json.loads(result)
+
+    deps = []
+
+    def get_deps(obj: Any):
+        for file in obj:
+            if isinstance(file, dict) and "location" in file:
+                loc = file.get("location")
+
+                # Check directory
+                if file.get("class") == "Directory":
+                    if file.get("listing"):
+                        get_deps(file.get("listing"))
+                    else:
+                        # no listing, so import all files in the directory
+                        for folder, _, sub_files in os.walk(loc):
+                            for sub_file in sub_files:
+                                deps.append(os.path.join(folder, sub_file))
+                else:
+                    deps.append(file.get("location"))
+
+                # check secondaryFiles
+                if "secondaryFiles" in file:
+                    get_deps(file.get("secondaryFiles"))
+
+    get_deps(json_result.get("secondaryFiles", []))
+    return deps
+
+
+def submit_run(client: WESClient,
+               cwl_file: str,
+               input_file: Optional[str] = None,
+               engine_options: Optional[List[str]] = None) -> str:
+    # First, get the list of files to attach to this workflow
+    attachments = get_deps_from_cwltool(cwl_file, input_file)
+
+    if input_file:
+        attachments.extend(get_deps_from_cwltool(cwl_file, input_file, input_deps=True))
+
+    # logging.warning(f"Files to import: {attachments}")
+
+    run_result: Dict[str, Any] = client.run(cwl_file, input_file,
+                                            attachments=attachments,
+                                            engine_options=engine_options)
+    return run_result.get("run_id", None)
+
+
+def poll_run(client: WESClient, run_id: str) -> bool:
+    """ Return True if the given workflow run is in a finished state."""
+    status_result = client.get_run_status(run_id)
+    state = status_result.get("state")
+
+    return state in ("COMPLETE", "CANCELING", "CANCELED", "EXECUTOR_ERROR", "SYSTEM_ERROR")
+
+
+def print_logs_and_exit(client: WESClient, run_id: str) -> None:
+    # cwltest checks for stdout, stderr, and return code
+    data = client.get_run_log(run_id)
```

Comment:
It doesn't look like this makes more requests to actually download the stdout and stderr data form the URLs that the WES server is supposed to give.

Reply:
Ah, I'm using the `outputs` data from the GetRunLog response here, since it's the same as the stdout for `toil-cwl-runner`.